[Inductor Cutlass backend] GEMM size threshold for Cutlass backend usage

kadeng opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #121491
* #121490

 * Adds a configurable GEMM size threshold for the usage of Cutlass GEMM Kernels **_inductor.config.cutlass_backend_min_gemm_size**

 * During GEMM algorithm choice generation: **if no viable choices can be generated using the configured backends, the ATen backend will be used as a fallback backend**, even if it is not enabled in **_inductor.config.max_autotune_gemm_backends**

Test plan:
CI
Additional unit test in test_cutlass_backend.py

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @muchulee8 @aakhundov @ColinPeppler @amjames @desertfire @chauhang

kadeng committed:
[Inductor Cutlass backend] GEMM size threshold for Cutlass backend usage

Adds a configurable GEMM size threshold for the usage of Cutlass GEMM Kernels

Test plan:
CI
Additional unit test in test_cutlass_backend.py

[ghstack-poisoned]

kadeng committed:
Update on "[Inductor Cutlass backend] GEMM size threshold for Cutlass backend usage"


Adds a configurable GEMM size threshold for the usage of Cutlass GEMM Kernels

Test plan:
CI
Additional unit test in test_cutlass_backend.py

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

kadeng committed:
Update on "[Inductor Cutlass backend] GEMM size threshold for Cutlass backend usage"


Adds a configurable GEMM size threshold for the usage of Cutlass GEMM Kernels

Test plan:
CI
Additional unit test in test_cutlass_backend.py

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

kadeng committed:
Update on "[Inductor Cutlass backend] GEMM size threshold for Cutlass backend usage"


Adds a configurable GEMM size threshold for the usage of Cutlass GEMM Kernels

Test plan:
CI
Additional unit test in test_cutlass_backend.py

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

kadeng committed:
Update on "[Inductor Cutlass backend] GEMM size threshold for Cutlass backend usage"


 * Adds a configurable GEMM size threshold for the usage of Cutlass GEMM Kernels **_inductor.config.cutlass_backend_min_gemm_size**

 * During GEMM algorithm choice generation: **if no viable choices can be generated using the configured backends, the ATen backend will be used as a fallback backend**, even if it is not enabled in **_inductor.config.max_autotune_gemm_backends**

Test plan:
CI
Additional unit test in test_cutlass_backend.py

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

kadeng committed:
Update on "[Inductor Cutlass backend] GEMM size threshold for Cutlass backend usage"


 * Adds a configurable GEMM size threshold for the usage of Cutlass GEMM Kernels **_inductor.config.cutlass_backend_min_gemm_size**

 * During GEMM algorithm choice generation: **if no viable choices can be generated using the configured backends, the ATen backend will be used as a fallback backend**, even if it is not enabled in **_inductor.config.max_autotune_gemm_backends**

Test plan:
CI
Additional unit test in test_cutlass_backend.py

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

kadeng committed:
Update on "[Inductor Cutlass backend] GEMM size threshold for Cutlass backend usage"


 * Adds a configurable GEMM size threshold for the usage of Cutlass GEMM Kernels **_inductor.config.cutlass_backend_min_gemm_size**

 * During GEMM algorithm choice generation: **if no viable choices can be generated using the configured backends, the ATen backend will be used as a fallback backend**, even if it is not enabled in **_inductor.config.max_autotune_gemm_backends**

Test plan:
CI
Additional unit test in test_cutlass_backend.py

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

jansel commented on torch/_inductor/utils.py:
```diff
@@ -977,7 +977,15 @@ def use_triton_template(layout, *, enable_int32=False):
-def use_cutlass_template(layout):
+def use_cutlass_template(layout, m, n, k):
+    gemm_size = -1
+    try:
+        # this fails if it's dynamic
+        gemm_size = int(m * n * k)
+    except Exception:
+        pass
```
Rather than erroring you can use `V.graph.sizevars.size_hint`.

kadeng replied:
Changed that here. I think I should also use it in the PR above it in this stack ( about tolerating dynamic shapes ) to determine GEMM dimensions.

kadeng committed:
Update on "[Inductor Cutlass backend] GEMM size threshold for Cutlass backend usage"


 * Adds a configurable GEMM size threshold for the usage of Cutlass GEMM Kernels **_inductor.config.cutlass_backend_min_gemm_size**

 * During GEMM algorithm choice generation: **if no viable choices can be generated using the configured backends, the ATen backend will be used as a fallback backend**, even if it is not enabled in **_inductor.config.max_autotune_gemm_backends**

Test plan:
CI
Additional unit test in test_cutlass_backend.py

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

kadeng committed:
Update on "[Inductor Cutlass backend] GEMM size threshold for Cutlass backend usage"


 * Adds a configurable GEMM size threshold for the usage of Cutlass GEMM Kernels **_inductor.config.cutlass_backend_min_gemm_size**

 * During GEMM algorithm choice generation: **if no viable choices can be generated using the configured backends, the ATen backend will be used as a fallback backend**, even if it is not enabled in **_inductor.config.max_autotune_gemm_backends**

Test plan:
CI
Additional unit test in test_cutlass_backend.py

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

jansel reviewed (changes requested):
Tests are failing

kadeng committed:
Update on "[Inductor Cutlass backend] GEMM size threshold for Cutlass backend usage"


 * Adds a configurable GEMM size threshold for the usage of Cutlass GEMM Kernels **_inductor.config.cutlass_backend_min_gemm_size**

 * During GEMM algorithm choice generation: **if no viable choices can be generated using the configured backends, the ATen backend will be used as a fallback backend**, even if it is not enabled in **_inductor.config.max_autotune_gemm_backends**

Test plan:
CI
Additional unit test in test_cutlass_backend.py

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

kadeng committed:
Update on "[Inductor Cutlass backend] GEMM size threshold for Cutlass backend usage"


 * Adds a configurable GEMM size threshold for the usage of Cutlass GEMM Kernels **_inductor.config.cutlass_backend_min_gemm_size**

 * During GEMM algorithm choice generation: **if no viable choices can be generated using the configured backends, the ATen backend will be used as a fallback backend**, even if it is not enabled in **_inductor.config.max_autotune_gemm_backends**

Test plan:
CI
Additional unit test in test_cutlass_backend.py

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

kadeng commented:
> Tests are failing

Yes, I have been making the changes and introduced a circular import. That one's fixed and I'm waiting for CI.

jansel commented on torch/_inductor/utils.py:
```diff
@@ -977,7 +977,17 @@ def use_triton_template(layout, *, enable_int32=False):
+    try:
+        # this fails if it's dynamic
+        gemm_size = V.graph.sizevars.size_hint(m * n * k)
+    except Exception:
```
Let's make this exception more narrow (if this is still needed).  This could hide bugs.

kadeng replied:
makes sense, I eliminated the try catch entirely..

kadeng committed:
Update on "[Inductor Cutlass backend] GEMM size threshold for Cutlass backend usage"


 * Adds a configurable GEMM size threshold for the usage of Cutlass GEMM Kernels **_inductor.config.cutlass_backend_min_gemm_size**

 * During GEMM algorithm choice generation: **if no viable choices can be generated using the configured backends, the ATen backend will be used as a fallback backend**, even if it is not enabled in **_inductor.config.max_autotune_gemm_backends**

Test plan:
CI
Additional unit test in test_cutlass_backend.py

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

kadeng committed:
Rebased on top of origin/viable/strict again on "[Inductor Cutlass backend] GEMM size threshold for Cutlass backend usage"


 * Adds a configurable GEMM size threshold for the usage of Cutlass GEMM Kernels **_inductor.config.cutlass_backend_min_gemm_size**

 * During GEMM algorithm choice generation: **if no viable choices can be generated using the configured backends, the ATen backend will be used as a fallback backend**, even if it is not enabled in **_inductor.config.max_autotune_gemm_backends**

Test plan:
CI
Additional unit test in test_cutlass_backend.py

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

kadeng commented:
Rebase introduced some new test failures, working on that..

kadeng committed:
Update

[ghstack-poisoned]

kadeng commented:
@pytorchbot merge

kadeng committed:
Skipping tests on rocm

[ghstack-poisoned]

kadeng committed:
Update

[ghstack-poisoned]

kadeng committed:
Update

[ghstack-poisoned]

kadeng commented:
@pytorchbot merge --ignore-current "Current test failures are unrelated ( see parent commit in stack)"

kadeng commented:
@pytorchbot merge --ignore-current -m "Current test failures are unrelated ( see parent commit in stack)"

kadeng commented:
@pytorchbot merge --ignore-current
